runtime/drivers/duckdb/connectors.go

```diff
@@ -123,14 +156,158 @@ func (c *connection) ingestLocalFiles(ctx context.Context, env *connectors.Env,
 	return &drivers.IngestionSummary{BytesIngested: bytesIngested}, nil
 }
 
-func fileSize(paths []string) int64 {
-	var size int64
-	for _, path := range paths {
-		if info, err := os.Stat(path); err == nil { // ignoring error since only error possible is *PathError
-			size += info.Size()
+func (c *connection) scanSchemaFromQuery(ctx context.Context, qry string) (map[string]string, error) {
+	result, err := c.Execute(ctx, &drivers.Statement{Query: qry, Priority: 1})
+	if err != nil {
+		return nil, err
+	}
+	defer result.Close()
+
+	schema := make(map[string]string)
+	for i := 0; result.Next(); i++ {
+		var s duckDBTableSchemaResult
+		if err := result.StructScan(&s); err != nil {
+			return nil, err
 		}
+		schema[s.ColumnName] = s.ColumnType
 	}
-	return size
+	return schema, nil
+}
+
+type appender struct {
+	*connection
+	source             *connectors.Source
+	ingestionProps     map[string]any
+	allowColAddition   bool
+	allowColRelaxation bool
+	tableSchema        map[string]string
+}
+
+func newAppender(c *connection, source *connectors.Source, ingestionProps map[string]any) (*appender, error) {
+	// parse required properties from source.Properties
+	allowColAddition, allowColRelaxation, err := schemaRelaxationProperties(source.Properties)
+	if err != nil {
+		return nil, err
+	}
+	return &appender{
+		connection:         c,
+		source:             source,
+		ingestionProps:     ingestionProps,
+		allowColAddition:   allowColAddition,
+		allowColRelaxation: allowColRelaxation,
+		tableSchema:        nil,
+	}, nil
+}
+
+func (a *appender) appendData(ctx context.Context, files []string, format string) error {
+	from, err := sourceReader(files, format, a.ingestionProps)
+	if err != nil {
+		return err
+	}
+
+	query := fmt.Sprintf("INSERT INTO %q (SELECT * FROM %s);", a.source.Name, from)
+	err = a.Exec(ctx, &drivers.Statement{Query: query, Priority: 1})
+	if err == nil || !containsAny(err.Error(), []string{"binder error", "conversion error"}) {
+		return err
+	}
```

Comment:
I might be missing something, but after the first insertion error that leads to relaxation, won't this always lead to an error? I think we can solve that by using `INSERT INTO %q BY NAME (SELECT * FROM %s)` – that would also enable support for column reordering in partitions. Does that sound right?

Reply:
Would it solve it if we only add `BY NAME` if `allow_field_relaxation` is true? By a slight stretch, you could argue that column reordering is also a kind of relaxation